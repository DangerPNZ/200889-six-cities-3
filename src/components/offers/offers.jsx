import React from 'react';
import PropTypes from 'prop-types';
import {withBooleanToggle} from '../../hocs/with-boolean-toggle/with-boolean-toggle.jsx';
import {OffersSortOptions} from '../offers-sort-options/offers-sort-options.jsx';
import OffersList from '../offers-list/offers-list.jsx';
import {Map} from '../map/map.jsx';

const OffersComponent = ({sortedOffers, offersSortType, onSortOptionClick, offerInMouseEnterId, selectedCity}) => {
  const SortOptionsWithVisibilityControl = withBooleanToggle(OffersSortOptions);
  return (<div className="cities">
    <div className="cities__places-container container">
      <section className="cities__places places">
        <h2 className="visually-hidden">Places</h2>
        <b className="places__found">{sortedOffers.length} places to stay in {selectedCity}</b>
        {
          <SortOptionsWithVisibilityControl
            offersSortType = {offersSortType}
            onSortOptionClick = {onSortOptionClick}
          />
        }
        <OffersList/>
      </section>
      <div className="cities__right-section">
        <section className="cities__map map">
          <Map
            offers = {sortedOffers}
            selectedOfferId = {offerInMouseEnterId}
          />
        </section>
      </div>
    </div>
  </div>
  );
};

OffersComponent.propTypes = {
  sortedOffers: PropTypes.arrayOf(PropTypes.exact({
    city: PropTypes.exact({
      name: PropTypes.string.isRequired,
      coordinates: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
      mapZoom: PropTypes.number.isRequired
    }).isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    goods: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    bedrooms: PropTypes.number.isRequired,
    host: PropTypes.exact({
      avatarUrl: PropTypes.string.isRequired,
      id: PropTypes.number.isRequired,
      isPro: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired
    }).isRequired,
    images: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    previewImage: PropTypes.string.isRequired,
    location: PropTypes.exact({
      coordinates: PropTypes.arrayOf(PropTypes.number.isRequired).isRequired,
      zoom: PropTypes.number.isRequired
    }).isRequired,
    id: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    type: PropTypes.string.isRequired,
    premium: PropTypes.bool.isRequired,
    isFavorites: PropTypes.bool.isRequired,
    rating: PropTypes.number.isRequired,
    maxAdults: PropTypes.number.isRequired
  }).isRequired).isRequired,

  offersSortType: PropTypes.string.isRequired,

  onSortOptionClick: PropTypes.func.isRequired,

  offerInMouseEnterId: PropTypes.number,

  selectedCity: PropTypes.string.isRequired
};

export const Offers = React.memo(OffersComponent);
