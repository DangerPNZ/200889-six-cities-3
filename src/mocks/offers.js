export const DataValue = {
  OFFERS_AMOUNT: `555`,
  OFFERS_NAMES: [
    `Beautiful & luxurious apartment at great location`,
    `Wood and stone place`,
    `Canal View Prinsengracht`,
    `Nice, cozy, warm big bed apartment`
  ],
  COORDINATES: [
    [52.3909553943508, 4.85309666406198],
    [52.369553943508, 4.85309666406198],
    [52.3909553943508, 4.929309666406198],
    [52.3809553943508, 4.939309666406198]
  ],
  REVIEWS: [
    {},
    {},
    {}
  ]
};
