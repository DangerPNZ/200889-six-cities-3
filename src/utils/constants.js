export const CITIES_FAULT_TOLERANT = [`Paris`, `Cologne`, `Brussels`, `Amsterdam`, `Hamburg`, `Dusseldorf`];
export const SortOption = {
  DEFAULT: `Popular`,
  BY_PRICE_LOW_TO_HIGHT: `Price: low to high`,
  BY_PRICE_HIGHT_TO_LOW: `Price: high to low`,
  BY_RATING_HIGHT_TO_LOW: `Top rated first`,
};
export const CompareDirection = {
  ASC: `ASC`,
  DESC: `DESK`
};
export const CardRenderMode = {
  MAIN: `MAIN`,
  NEAR: `NEAR`,
  FAVORITE: `FAVORITE`
};
export const ReducerName = {
  FETCHED_DATA: `FETCHED_DATA`,
  CONTEXT: `CONTEXT`,
  USER: `USER`
};
export const AuthorizationStatus = {
  NO_AUTH: `NO_AUTH`,
  AUTHORIZED: `AUTHORIZED`
};
export const PagePath = {
  MAIN: `/`,
  OFFER: `/offer/`,
  LOGIN: `/login`,
  FAVORITES: `/favorites`
};
