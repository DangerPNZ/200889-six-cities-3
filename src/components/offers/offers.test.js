import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from "redux-mock-store";
import {BrowserRouter} from 'react-router-dom';
import {Offers} from './offers.jsx';
import {Provider} from 'react-redux';
import {CITIES_FAULT_TOLERANT, SortOption, ReducerName, AuthorizationStatus} from '../../utils/constants.js';

const mockStore = configureStore([]);
const TestDataValue = {
  OFFERS: [
    {
      city: {
        name: `Hamburg`,
        coordinates: [0],
        mapZoom: 10
      },
      name: `Amazing room`,
      description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.`,
      goods: [`Lorem`, `ipsum`, `dolor`, `sit amet`, `consectetur`],
      bedrooms: 3,
      host: {
        avatarUrl: `./img.jpg`,
        id: 5,
        isPro: true,
        name: `Samantha`
      },
      images: [`./photo.jpg`, `./photo_1.jpg`, `./photo_2.jpg`, `./photo_3.jpg`],
      previewImage: `./preview.jpg`,
      location: {
        coordinates: [0],
        zoom: 10
      },
      id: 1,
      price: 100,
      type: `room`,
      premium: true,
      isFavorites: false,
      rating: 4,
      maxAdults: 1
    },
    {
      city: {
        name: `Hamburg`,
        coordinates: [0],
        mapZoom: 10
      },
      name: `Light room`,
      description: `Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.`,
      goods: [`Lorem`, `ipsum`, `dolor`, `sit amet`, `consectetur`],
      bedrooms: 2,
      host: {
        avatarUrl: `./img.jpg`,
        id: 5,
        isPro: true,
        name: `Samantha`
      },
      images: [`./photo.jpg`, `./photo_1.jpg`, `./photo_2.jpg`, `./photo_3.jpg`],
      previewImage: `./preview.jpg`,
      location: {
        coordinates: [0],
        zoom: 10
      },
      id: 2,
      price: 150,
      type: `room`,
      premium: false,
      isFavorites: true,
      rating: 4,
      maxAdults: 1
    }
  ],
  OFFERS_SORT_TYPE: `Price: low to high`
};

it(`Offers component structure test`, () => {
  const store = mockStore({
    [ReducerName.FETCHED_DATA]: {
      offers: TestDataValue.OFFERS
    },
    [ReducerName.CONTEXT]: {
      selectedCity: CITIES_FAULT_TOLERANT[0],
      offersSortType: SortOption.DEFAULT
    },
    [ReducerName.USER]: {
      authorizationStatus: AuthorizationStatus.AUTHORIZED
    }
  });
  const tree = renderer
  .create(
      <BrowserRouter>
        <Provider store = {store}>
          <Offers
            sortedOffers = {TestDataValue.OFFERS}
            offersSortType = {TestDataValue.OFFERS_SORT_TYPE}
            selectedCity = {CITIES_FAULT_TOLERANT[0]}
            onSortOptionClick = {() => {}}
          />
        </Provider>
      </BrowserRouter>,
      {
        createNodeMock: () => {
          return {};
        }
      }
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
