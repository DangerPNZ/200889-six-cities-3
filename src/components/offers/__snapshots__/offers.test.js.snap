// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Offers component structure test 1`] = `
<div
  className="cities"
>
  <div
    className="cities__places-container container"
  >
    <section
      className="cities__places places"
    >
      <h2
        className="visually-hidden"
      >
        Places
      </h2>
      <b
        className="places__found"
      >
        2
         places to stay in 
        Paris
      </b>
      <form
        action="#"
        className="places__sorting"
        method="get"
      >
        <span
          className="places__sorting-caption"
        >
          Sort by
        </span>
        <span
          className="places__sorting-type"
          onClick={[Function]}
          tabIndex="0"
        >
          Price: low to high
          <svg
            className="places__sorting-arrow"
            height="4"
            width="7"
          >
            <use
              xlinkHref="#icon-arrow-select"
            />
          </svg>
        </span>
        <ul
          className="places__options places__options--custom"
        >
          <li
            className="places__option"
            onClick={[Function]}
            tabIndex="0"
          >
            Popular
          </li>
          <li
            className="places__option places__option--active"
            onClick={[Function]}
            tabIndex="0"
          >
            Price: low to high
          </li>
          <li
            className="places__option"
            onClick={[Function]}
            tabIndex="0"
          >
            Price: high to low
          </li>
          <li
            className="places__option"
            onClick={[Function]}
            tabIndex="0"
          >
            Top rated first
          </li>
        </ul>
      </form>
      <div
        className="cities__places-list places__list tabs__content"
      />
    </section>
    <div
      className="cities__right-section"
    >
      <section
        className="cities__map map"
      >
        <div
          id="map"
          style={
            Object {
              "height": "100%",
            }
          }
        />
      </section>
    </div>
  </div>
</div>
`;
